docs/proposal/20210823-gpu-support.md

```diff
@@ -0,0 +1,197 @@
+}
+
+//VGPUDeviceSpec defines virtual machine's VGPU configuration
+type VGPUDeviceSpec struct {
+ // ProfileName is the Profile Name of a virtual machine's VGPU, in string.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ ProfileName string `json:"profileName,omitempty"`
+}
+
+//PCIDeviceSpec defines virtual machine's PCI configuration
+type PCIDeviceSpec struct {
+
+ // DeviceId is the device ID of a virtual machine's PCI, in integer.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ DeviceId int32 `json:"deviceId,omitempty"`
+
+ // VendorId is the vendor ID of a virtual machine's PCI, in integer.
+ // Defaults to the eponymous property value in the template from which the
+ // virtual machine is cloned.
+ // +optional
+ VendorId int32 `json:"vendorId,omitempty"`
+}
+
+```
+
+#### Controller Changes
+
+- If VGPU is present, VGPU's `ProfileName` would be fetched from `VirtualMachineCloneSpec` and device object (`createPCIPassThroughDevice` from govmomi library) is formed using this VGPU `ProfileName`. This device object is added to
+- the slice of devices and then used by `vspheremachine_controller` to schedule nodes.
+- If PCI device is present, PCI device's `DeviceId` and `VendorId` would be fetched from `VirtualMachineCloneSpec` and device object(`createPCIPassThroughDevice` from govmomi library) is formed using this information and then used by the `vspheremachine_controller` to schedule nodes.
+- The controller initiallizes variables respective to VGPU i.e `GPUProfileName`.
+- Similarly, controller also initializes variables related to Direct devices(for PCI passthrough and GPUDirect) ie. `GPUDeviceId` and `GPUVendorId`.
+- Using the above variables, respective objects of backing info of type `vimTypes.VirtualPCIPassthroughVmiopBackingInfo` with `{Vgpu: <GPUProfileName>}` for VGPU's and `vimTypes.VirtualPCIPassthroughAllowedDevice` for GPUDirect should be created by the controller.
+- Controller will call a function called as createPCI device which will use the information in above step to create objects of type `vimTypes.VirtualPCIPassthrough`
+- The create PCI device function will also be responsible for comparing the current devices with the expected devices.
+- The create method will return a slice `deviceChanges` of type `vimTypes.BaseVirtualDeviceConfigSpec`.
+- If there are no current devices and new devices are added. The slice `devicechanges` would reflect only add operations which would be equivalent to the number of devices added.
+- If the current devices are same as proposed devices. The length of the device changes remains 0.
+- Adding Vgpu devices and PCI devices with differnet backing info would return one remove and one add operation for each device.
+So the length of device changes would be 2*devices with different backing info.
```

Comment:
Since the VSphereMachineTemplate and VSphereMachine objects are immutable, how would this case arise? Why would vGPU or PCI devices with different backing info be needed?

Reply:
Just curious, how does Network Devices work. What I see is, they call a `Finder.Network` in `govmomi` and just return that list of devices. In the case of GPU's there is no such method as `finder.PCIDevices`. All the workflow above will not be required if we have a list of GPU devices. 

P.S If above stands true, then I guess, I have to implement a PCI finder in govmomi.

Reply:
//need to file an issue in govmomi

Reply:
This seems to be taken care of by https://github.com/vmware/govmomi/pull/2772. Resolving this